xls/modules/rle/rle_dec.x

```diff
@@ -0,0 +1,520 @@
+// (set state value)    ╨─────╨─────╨─────╨───────────╨─────╨─────╨─────╨────
+//
+//                      ──────╥───────────╥───────────╥─────────────────╥────
+// symbol               XXXXXX║ A         ║ B         ║ C               ║XXXX
+// (output channel)     ──────╨───────────╨───────────╨─────────────────╨────
+// last                                   ┌─────┐                 ┌─────┐
+// (output channel)     ──────────────────┘     └─────────────────┘     └────
+
+
+import std
+import xls.modules.rle.rle_common as rle_common
+
+type DecInData  = rle_common::EncData;
+type DecOutData = rle_common::PlainData;
+
+// structure to preserve the state of an RLE decoder
+struct RunLengthDecoderState<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+  // symbol to be repeated on output
+  symbol: bits[SYMBOL_WIDTH],
+  // count of symbols that has to be send
+  count: bits[COUNT_WIDTH],
+  // send last when repeat ends
+  last: bool,
+}
+ // RLE decoder implementation
+pub proc RunLengthDecoder<SYMBOL_WIDTH: u32, COUNT_WIDTH: u32> {
+  input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in;
+  output_s: chan<DecOutData<SYMBOL_WIDTH>> out;
+
+  init {(
+    RunLengthDecoderState<SYMBOL_WIDTH, COUNT_WIDTH> {
+      symbol: bits[SYMBOL_WIDTH]:0,
+      count:  bits[COUNT_WIDTH]:0,
+      last:   bool:false,
+    }
+  )}
+
+  config (
+    input_r: chan<DecInData<SYMBOL_WIDTH, COUNT_WIDTH>> in,
+    output_s: chan<DecOutData<SYMBOL_WIDTH>> out,
+  ) {(input_r, output_s)}
+
+  next (tok: token, state: RunLengthDecoderState<SYMBOL_WIDTH, COUNT_WIDTH>) {
+    let zero_input = DecInData {
+      symbol: bits[SYMBOL_WIDTH]:0,
+      count: bits[COUNT_WIDTH]:0,
+      last: false
+    };
+    let recv_next_symbol = (state.count == bits[COUNT_WIDTH]:0);
+    let (tok, input) = recv_if(tok, input_r, recv_next_symbol, zero_input);
+    let (next_symbol, count, next_last) = if (recv_next_symbol) {
+      (input.symbol, input.count, input.last)
+    } else {
+      (state.symbol, state.count, state.last)
+    };
+    let next_count = count - bits[COUNT_WIDTH]:1;
+    let done_sending = (next_count == bits[COUNT_WIDTH]:0);
+    let send_last = next_last && done_sending;
+    let data_tok = send(tok, output_s, DecOutData {
+      symbol: next_symbol,
+      last: send_last
+    });
+    if (send_last) {
+      zero!<RunLengthDecoderState>()
+    } else {
+      RunLengthDecoderState {
+        symbol: next_symbol,
+        count: next_count,
+        last: next_last,
+      }
+    }
+  }
+}
+
+
+// RLE decoder specialization for the codegen
+proc RunLengthDecoder32 {
+  init {()}
+
+  config (
+    input_r: chan<DecInData<32, 2>> in,
+    output_s: chan<DecOutData<32>> out,
+  ) {
+    spawn RunLengthDecoder<u32:32, u32:2>(input_r, output_s);
+    ()
+  }
+
+  next (tok: token, state: ()) {
+    ()
+  }
+}
+
+// Tests
+
+const TEST_SYMBOL_WIDTH = u32:32;
+const TEST_COUNT_WIDTH  = u32:32;
+
+type TestCommonSymbol     = bits[TEST_SYMBOL_WIDTH];
```

Comment:
nit: should this be named after the test `SingleSymbol`?

Reply:
or simply drop the `Command` if it's shared between multiple tests.

Reply:
But it's `Common`, so maybe it can be left as it is, wdyt @proppy?

Reply:
yep, I understand it's being shared by the tests but I think the definition appearing at the beginning of the file is enough of a hint to convey that it's being shared (without having to prefix all the name by `Common`)

Reply:
I've removed `Common` from naming.